### PR TITLE
Issue 7623  - Heap Buffer Overflow in 389-ds-base Audit Log Password Masking

### PR DESCRIPTION
**Pull request by mreynolds389**
Description:

When password storage scheme is set to CLEAR and you set a password that is less than 23-bytes then a possible heap overflow is possible.

Replace the unsafe `strcpy` with a bounded copy that respects the available space in the buffer

Severity: Low (CVSS 3.3)
CWE: CWE-122 (Heap-based Buffer Overflow)

relates: https://github.com/389ds/389-ds-base/issues/7623

## Summary by Sourcery

Prevent heap buffer overflow in audit log password masking and add regression coverage for ASAN-detected overflows.

Bug Fixes:
- Fix potential heap-based buffer overflow when masking short CLEAR passwords in audit log entries.

Tests:
- Add ASAN-gated audit log test to ensure password masking does not cause heap-buffer-overflow with short CLEAR passwords.

### CODE REVIEW

File: ldap/servers/slapd/auditlog.c [hunk 1]
```diff
@@ -106,7 +106,25 @@ create_masked_entry_string(Slapi_Entry *original_entry, int *len)
...
+                /* Calculate available space from colon_pos+1 to end of value */
+                if (next_line != NULL) {
+                    avail = (size_t)(next_line - (colon_pos + 1));
+                } else {
+                    avail = strlen(colon_pos + 1);
+                }
+
+                if (mask_len <= avail) {
+                    memcpy(colon_pos + 1, mask, mask_len);
+                    /* NUL-fill remaining space to avoid leaking original value */
```
Comment (progier389) [inline ldap/servers/slapd/auditlog.c]: Comment should be: Space-fill instead of NUL-fill

File: ldap/servers/slapd/auditlog.c [hunk 2]
```diff
@@ -106,7 +106,25 @@ create_masked_entry_string(Slapi_Entry *original_entry, int *len)
 
             /* Check if this is a password attribute that needs masking */
             if (is_password_attribute(line_start)) {
-                strcpy(colon_pos + 1, " **********************");
+                const char *mask = " **********************";
```
Comment (progier389) [inline ldap/servers/slapd/auditlog.c]: I wonder if the following would not be better:
```
  const char maskj[] = " **********************";
  size_t mask_len = (sizeof mask)-1;
```
( so the size get computed by the compiler rather than by strlen  )

**Review by progier389 — APPROVED** [review]
Looks good except 2 minor points
